components/video/video.tsx

```diff
@@ -0,0 +1,319 @@
+      }
+    }
+    const onLoadedMetadata = () => {
+      if (videoRef.current) setDuration(videoRef.current.duration)
+    }
+    const play = () => {
+      if (videoRef.current) videoRef.current.play()
+    }
+    const pause = () => {
+      if (videoRef.current) videoRef.current.pause()
+    }
+    const toggle = () => {
+      if (isPlaying) return pause()
+      play()
+    }
+    const updateHandlePosition = (e: TouchEvent | MouseEvent) => {
+      let pageX
+      if (e instanceof TouchEvent) {
+        pageX = e.touches[0].pageX
+      } else {
+        pageX = e.pageX
+      }
+      if (dragHandlerRef.current) {
+        const rect = dragHandlerRef.current.getBoundingClientRect()
+        const position = ((pageX - (rect.left || 0)) / (rect.width || 0)) * 100
+        handlePosition.current = parseInt(`${Math.min(Math.max(position, 0), 100)}`) || 0
+      }
+    }
+    const startDrag = (e: MouseEvent | TouchEvent) => {
+      setDragging(true)
+      updateHandlePosition(e)
+    }
+    useEffect(() => {
+      const onDarg = (e: MouseEvent) => {
+        if (isDragging) updateHandlePosition(e)
+      }
+      const endDrag = () => {
+        setDragging(false)
+        if (videoRef.current) {
+          videoRef.current.currentTime = (handlePosition.current * videoRef.current.duration) / 100
+          setCurrentTime(videoRef.current.currentTime)
+        }
+      }
+      window.addEventListener('mousemove', onDarg)
+      window.addEventListener('touchmove', onDarg)
+      window.addEventListener('mouseup', endDrag)
+      window.addEventListener('touchend', endDrag)
+      return () => {
+        window.removeEventListener('mousemove', onDarg)
+        window.removeEventListener('touchmove', onDarg)
+        window.removeEventListener('mouseup', endDrag)
+        window.removeEventListener('touchend', endDrag)
+      }
+    }, [isDragging])
+
+    return (
+      <span className={className}>
+        <figure>
+          <main>
+            <div
+              className={`container ${controlsVisible ? 'controls-visible' : ''}`}
+              ref={containerRef}
+              onMouseMove={handleMouseMove}>
+              <video
+                onClick={toggle}
+                ref={videoRef}
+                onLoadedMetadata={onLoadedMetadata}
+                onTimeUpdate={handleProgress}
+                onPlay={() => setPlaying(true)}
+                onPause={() => setPlaying(false)}
+                src={src}
+                playsInline
+                {...props}
+              />
+              <div className="video-controls">
+                {isPlaying ? (
+                  <button onClick={pause} className="play-button">
+                    <PauseIcon size={14} />
+                  </button>
+                ) : (
+                  <button onClick={play} className="play-button">
+                    <PlayIcon size={14} />
+                  </button>
+                )}
```

Comment:
Maybe we can abstract this into a single component? The `video.tsx` file looks a little big.

```jsx
<VideoControlButton isPlaying={isPlaying} onPause={pause} onPlay={play} />
<VideoControlFullScreen isFullScreen={isFullScreen} onClick={fullScreenClick} />
```